### PR TITLE
VACMS-18003: Vet center services view

### PR DESCRIPTION
**Pull request by omahane**
## Description

Relates to #18003 

## Testing done
Manually.

## Screenshots
Screenshot 2024-07-17 at 10 27 21 AM

### Set up QA Content Publisher
- [x] As an admin, update the QA Content Publisher
  - [x] Assign the following roles
    - [x] Content creator - Vet Center
    - [x] Content publisher
  - [x] Assign the following Section
    - [x] Charleston, WV Vet Center
    - [x] Chicago Vet Center

### View services for one Vet Center facility
- [x] Log in as an QA Content Publisher
- [x] Click "Charleston, WV Vet Center" section
- [x] Click **Go to services** in the dashboard
- [x] Confirm that Charleston, WV Vet Center services are showing.
- [x] Confirm that the first column refers to the facility
- [x] Confirm that the column headings are correct
- [x] Confirm that the services are listed first by "required", then "optional" and finally alphabetically within each grouping.
- [x] Click on a service name
- [x] Confirm that service node opens in a new tab in node:view
- [x] Confirm that the service is for the Charleston, WV Vet Center
- [x] Close the service tab
- [x] Click on the **Edit** button.
- [x] Confirm that the service node opens in a new tab in node: edit
- [x] Confirm that the service is for the Charleston, WV Vet Center
- [x] Close the service tab

### View services for another Vet Center facility
- [x] Go back to the [user profile](https://pr18531-rutrbjgdty8li53aalhrxlpalospnanb.ci.cms.va.gov/user) page.
- [x] Click "Chicago Vet Center" section
- [x] Click **Go to services** in the dashboard
- [x] Confirm that Chicago Vet Center services are showing.
- [x] Click on a service name
- [x] Confirm that service node opens in a new tab in node:view
- [x] Confirm that the service is for the Chicago Vet Center
- [x] Close the service tab
- [x] Click on the **Edit** button.
- [x] Confirm that the service node opens in a new tab in node: edit
- [x] Confirm that the service is for the Chicago Vet Center
- [x] Close the service tab

### View front-end to check for regressions
- [x] Go back to the [user profile](https://pr18531-rutrbjgdty8li53aalhrxlpalospnanb.ci.cms.va.gov/user) page.
- [x] Click "Chicago Vet Center" section
- [x] Click **View** under "Main Vet Center page"
- [x] Open the [VA.gov URL](https://web-rutrbjgdty8li53aalhrxlpalospnanb.ci.cms.va.gov/chicago-vet-center/) in a new tab
- [x] Compare the services in Drupal to the services on the front-end
- [x] Confirm that they are the same

### Edit a required service
- [x] Go back to the [user profile](https://pr18531-rutrbjgdty8li53aalhrxlpalospnanb.ci.cms.va.gov/user) page.
- [x] Click "Chicago Vet Center" section
- [x] Click **Go to services** in the dashboard
- [x] Click on "Addiction and substance use care"
- [x] Edit the "Vet Center service description"
- [x] Confirm that the grey box under the Vet Center service description describes this as a required service
- [x] Choose "Published" under **Save as**
- [x] Confirm that "Published" and "Draft" are the only **Save as** options
- [x] Click **Save**
- [x] Close the tab

### Edit an optional service
- [x] On Chicago Vet Center, edit "Telehealth"
- [x] Confirm that the grey box under the Vet Center service description describes this as a optional service
- [x] Choose "Archived" under **Save as**
- [x] Click **Save**
- [x] Close the tab
- [x] On the Chicago Vet Center, refresh the page
- [x] Confirm that "Telehealth" shows as "Archived"

### Add a service
- [x] Click **Add a new service**
- [x] In the new tab, confirm that the **Section** is "Chicago Vet Center"
- [x] Confirm that the **Vet Center** is "Chicago Vet Center"
- [x] In the **Service name**, choose "Whole health"
- [x] Confirm that the service description populates the grey box under the **Service name**
- [x] In the **Service name**, choose "Workshops and classes"
- [x] Confirm that the service description populates with changed and correct information
- [x] Add a brief description under **Vet Center service description**
- [x] Choose "Draft" under **Save as** (which is the default)
- [x] Add a revision message
- [x] Save the service
- [x] Close the tab
- [x] On the Chicago Vet Center, refresh the page
- [x] Confirm that "Workshops and classes" shows as "Draft"

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [x] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

### Is this PR blocked by another PR?

- [x] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by laflannery** [comment]
I have realized that we need to warn folks that these links are going to open in a new tab but it would be insane to add "opens in a new tab" to every single one, especially the edit links. So I would like to edit the description text slightly. The below text was proposed and approved [in slack here](https://dsva.slack.com/archives/C0FQSS30V/p1721403451283619?thread_ts=1721321342.759269&cid=C0FQSS30V):

> Add or edit services that Veterans can receive at your Vet Center. Some content won't be editable because it comes from other sources. All links will open in a new tab. For full guidance, see How to add or edit a Vet Center service (opens in a new tab).

**Comment by laflannery** [comment]
LGTM!

**Review by dsasser — COMMENTED** [review]
I have provided code level comments, and I have some additional feedback that I'll put here:

Reuse:
Overall I like that the approach works and makes sense for this use case. However, reusability is going to be a challenge here. Since we targeting elements by name in [some places](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/18531/files#diff-918aa72ec728b8034464371325b30bfed0abd555c12f44af6bce1796ec839bd3R196),  replacing other IEF implementations would require a separate set of unique form alters, a viewsfield field, etc.

One idea I was tossing around was to build a field widget that would replace the current IEF field widget. Problem is that building such a widget is more work than what you have here, so there is that.

I did find a module, that you may have already evaluated, that does get close to what I was thinking, called [entity_reference_views](https://www.drupal.org/project/entity_reference_views). It's sorta a mashup between viewsfield and the work you have here. We wouldn't need everything it does (ajax), but it is possible this module could be patched, or used as a base for a custom module to suite our needs. And, maybe even could be contributed back if we wanted to go there.

That said, I'm not suggesting this specific thing is the only path forward. Maybe we get this in now and move onto the field widget later...or some other solution. But I think it warrants a conversation if we want to keep this as is, so we can as a team agree to accept the tech debt.

Tests:
None exist. Maybe none are needed?

File: docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module [hunk 1]
```diff
@@ -183,10 +184,58 @@ function va_gov_vet_center_vc_cap_address_alter_label_after_build(array $element
...
-    $form['#attached']['library'][] = 'va_gov_vet_center/set_ief_service_selects';
+  if ($entity->bundle() !== 'vet_center') {
+    return;
   }
+  $form['#attached']['library'][] = 'va_gov_vet_center/set_ief_service_selects';
+
+  // Attach the vet center services view to the form.
+  $form['group_vet_center_services']['vet_center_services'] = [
+    '#type' => 'view',
+    '#name' => 'vet_center_services',
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module:197]: Not only could I not get the config change alone to render the view in node:edit, but I found that I needed to hide it in node:view, anyway. Instead, I removed that Viewfield field totally and am attaching it in the alter only.

File: docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module [hunk 2]
```diff
@@ -183,10 +184,58 @@ function va_gov_vet_center_vc_cap_address_alter_label_after_build(array $element
...
+
+/**
+ * Implements hook_views_pre_render().
+ */
+function va_gov_vet_center_views_pre_render(ViewExecutable $view) {
+  // Check if this is the 'vet_center_services' view.
+  if ($view->id() !== 'vet_center_services') {
+    return;
+  }
+  $current_path = \Drupal::service('path.current')->getPath();
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module]: Great suggestion. Done.

File: config/sync/views.view.knowledge_base_search_results.yml [hunk 3]
```diff
@@ -258,7 +258,7 @@ display:
         type: none
         options: {  }
       cache:
-        type: none
+        type: search_api_none
```
[the remark this answers is not in the document]
Comment (omahane) [reply config/sync/views.view.knowledge_base_search_results.yml:261]: Thanks. Anything I can do to help.

File: docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module [hunk 4]
```diff
@@ -183,10 +184,58 @@ function va_gov_vet_center_vc_cap_address_alter_label_after_build(array $element
...
+  }
+  else {
+    return;
+  }
+  $entity = \Drupal::entityTypeManager()->getStorage('node')->load($node_id);
+  $entity_title = $entity->getTitle();
+
+  // Ensure the page title is not null and the desired field exists in the view.
+  if ($entity_title && isset($view->field['field_service_name_and_descripti'])) {
+    // Append or prepend the page title to the field label as needed.
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module]: OK, that's fixed.

File: docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module [hunk 5]
```diff
@@ -183,10 +184,58 @@ function va_gov_vet_center_vc_cap_address_alter_label_after_build(array $element
...
+    $new_label = "Services offered at $entity_title";
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vet_center/va_gov_vet_center.module]: Done.

**Comment by omahane** [comment]
> I have provided code level comments, and I have some additional feedback that I'll put here:
> 
> Reuse: Overall I like that the approach works and makes sense for this use case. However, reusability is going to be a challenge here. Since we targeting elements by name in [some places](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/18531/files#diff-918aa72ec728b8034464371325b30bfed0abd555c12f44af6bce1796ec839bd3R196), replacing other IEF implementations would require a separate set of unique form alters, a viewsfield field, etc.
> 
> One idea I was tossing around was to build a field widget that would replace the current IEF field widget. Problem is that building such a widget is more work than what you have here, so there is that.
> 
> I did find a module, that you may have already evaluated, that does get close to what I was thinking, called [entity_reference_views](https://www.drupal.org/project/entity_reference_views). It's sorta a mashup between viewsfield and the work you have here. We wouldn't need everything it does (ajax), but it is possible this module could be patched, or used as a base for a custom module to suite our needs. And, maybe even could be contributed back if we wanted to go there.
> 
> That said, I'm not suggesting this specific thing is the only path forward. Maybe we get this in now and move onto the field widget later...or some other solution. But I think it warrants a conversation if we want to keep this as is, so we can as a team agree to accept the tech debt.
> 
> Tests: None exist. Maybe none are needed?

I definitely will consider a widget approach next time. As far as I know IEF is completely out of the picture (at least in Facilities) with this removal. I took a look at the module and added it as a test. It would definitely need some work before we could use it.

## Tests
We could write a test in Cypress (sigh) that could look at a given Vet Center and check for a specific required service (checking the links to make sure they are going to the right place).

**Review by dsasser — APPROVED** [review]
Simple, effective solution; nice work!

I went through the QA steps and everything checked out as well, eg:

Screenshot 2024-07-24 at 1 34 53 PM

**Review by anantais — APPROVED** [review]

**Comment by omahane** [comment]
@jilladams Thanks for merging.

**Pull request merged**